cranelift/codegen/src/postopt.rs

```diff
@@ -360,20 +360,23 @@ fn optimize_complex_addresses(pos: &mut EncCursor, inst: Inst, isa: &dyn TargetI
 pub fn do_postopt(func: &mut Function, isa: &dyn TargetIsa) {
     let _tt = timing::postopt();
     let mut pos = EncCursor::new(func, isa);
+    let is_mach_backend = isa.get_mach_backend().is_some();
     while let Some(_block) = pos.next_block() {
         let mut last_flags_clobber = None;
         while let Some(inst) = pos.next_inst() {
             if isa.uses_cpu_flags() {
                 // Optimize instructions to make use of flags.
                 optimize_cpu_flags(&mut pos, inst, last_flags_clobber, isa);
 
-                // Track the most recent seen instruction that clobbers the flags.
-                if let Some(constraints) = isa
-                    .encoding_info()
-                    .operand_constraints(pos.func.encodings[inst])
-                {
-                    if constraints.clobbers_flags {
-                        last_flags_clobber = Some(inst)
+                if !is_mach_backend {
```

Comment:
The optimization here is converting a comparison opcode that produces an integer result and a branch that consumes an integer result into a comparison that produces a flags result and a branch that consumes the flags result. Currently it does this by rewriting the comparison in place, so there could be instructions between the compare and the branch. If they clobber the flags, the optimization isn't correct. So it isn't safe to disable this check. A possible alternative though would be to change the optimization to move the comparison just before the branch, to ensure that the flags value isn't clobbered.

Reply:
Ah, OK -- the arm64 backend at least will correctly handle an `icmp` + `brz`/`brnz` sequence, so we don't need this optimization -- I've just altered it to skip this mutation entirely when a MachBackend is in use.